File: app/src/main/java/education/mahmoud/quranyapp/data_layer/local/QuranDB.java
package education.mahmoud.quranyapp.data_layer.local;

import android.app.Application;

import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;

@Database(entities = {AyahItem.class, SuraItem.class}, version = 2, exportSchema = false)
public abstract class QuranDB extends RoomDatabase {

    // Singleton Pattern only one istance exists and availbale for all classes from this class
    private static QuranDB instance;

    public static synchronized QuranDB getInstance(Application application) {
        if (instance == null) { // first time to create instance
            instance = Room.databaseBuilder(application, QuranDB.class, "quran")
                    .fallbackToDestructiveMigration()
                    .allowMainThreadQueries()
                    .build();
        }
        return instance;
    }

    public abstract AyahDAO ayahDAO();

    public abstract SurahDAO surahDAO();
}

File: app/src/main/java/education/mahmoud/quranyapp/datalayer/local/room/QuranDB.kt
package education.mahmoud.quranyapp.datalayer.local.room

import androidx.room.Database
import androidx.room.RoomDatabase
import androidx.room.TypeConverters

@Database(entities = [AyahItem::class, SuraItem::class, BookmarkItem::class, ReadLog::class,
    RecordItem::class],
    version = 1,
    exportSchema = false)
@TypeConverters(PagesConverter::class)
abstract class QuranDB : RoomDatabase() {
    abstract fun ayahDAO(): AyahDAO
    abstract fun surahDAO(): SurahDAO
    abstract fun bookmarkDao(): BookmarkDAO
    abstract fun readLogDAO(): ReadLogDAO
    abstract fun recordDAO(): RecordDAO
}
